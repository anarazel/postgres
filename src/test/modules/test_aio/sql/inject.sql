SELECT count(*) FROM tbl_b WHERE ctid = '(2, 1)';

-- injected what we'd expect
SELECT inj_io_short_read_attach(8192);
SELECT invalidate_rel_block('tbl_b', 2);
SELECT count(*) FROM tbl_b WHERE ctid = '(2, 1)';
SELECT inj_io_short_read_detach();

-- injected a read shorter than a single block, expecting error
SELECT inj_io_short_read_attach(17);
SELECT invalidate_rel_block('tbl_b', 2);
SELECT redact($$
  SELECT count(*) FROM tbl_b WHERE ctid = '(2, 1)';
$$);
SELECT inj_io_short_read_detach();

-- shorten multi-block read to a single block, should retry, but that's not
-- implemented yet
SELECT inj_io_short_read_attach(8192);
SELECT invalidate_rel_block('tbl_b', 0);
SELECT invalidate_rel_block('tbl_b', 1);
SELECT invalidate_rel_block('tbl_b', 2);
SELECT redact($$
  SELECT count(*) FROM tbl_b;
$$);
SELECT inj_io_short_read_detach();

-- verify that checksum errors are detected even as part of a shortened
-- multi-block read
-- (tbl_a, block 1 is corrupted)
SELECT redact($$
  SELECT count(*) FROM tbl_a WHERE ctid < '(2, 1)';
$$);
SELECT inj_io_short_read_attach(8192);
SELECT invalidate_rel_block('tbl_a', 0);
SELECT invalidate_rel_block('tbl_a', 1);
SELECT invalidate_rel_block('tbl_a', 2);
SELECT redact($$
  SELECT count(*) FROM tbl_a WHERE ctid < '(2, 1)';
$$);
SELECT inj_io_short_read_detach();


-- FIXME: Should error
-- FIXME: errno encoding?
SELECT inj_io_short_read_attach(-5);
SELECT invalidate_rel_block('tbl_b', 2);
SELECT redact($$
  SELECT count(*) FROM tbl_b WHERE ctid = '(2, 1)';
$$);
SELECT inj_io_short_read_detach();
