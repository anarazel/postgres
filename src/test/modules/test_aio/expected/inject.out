SELECT count(*) FROM tbl_b WHERE ctid = '(2, 1)';
 count 
-------
     1
(1 row)

-- injected what we'd expect
SELECT inj_io_short_read_attach(8192);
 inj_io_short_read_attach 
--------------------------
 
(1 row)

SELECT invalidate_rel_block('tbl_b', 2);
 invalidate_rel_block 
----------------------
 
(1 row)

SELECT count(*) FROM tbl_b WHERE ctid = '(2, 1)';
 count 
-------
     1
(1 row)

SELECT inj_io_short_read_detach();
 inj_io_short_read_detach 
--------------------------
 
(1 row)

-- injected a read shorter than a single block, expecting error
SELECT inj_io_short_read_attach(17);
 inj_io_short_read_attach 
--------------------------
 
(1 row)

SELECT invalidate_rel_block('tbl_b', 2);
 invalidate_rel_block 
----------------------
 
(1 row)

SELECT redact($$
  SELECT count(*) FROM tbl_b WHERE ctid = '(2, 1)';
$$);
NOTICE:  wrapped error: could not read blocks 2..2 in file base/<redacted>: read only 0 of 8192 bytes
 redact 
--------
 f
(1 row)

SELECT inj_io_short_read_detach();
 inj_io_short_read_detach 
--------------------------
 
(1 row)

-- shorten multi-block read to a single block, should retry, but that's not
-- implemented yet
SELECT inj_io_short_read_attach(8192);
 inj_io_short_read_attach 
--------------------------
 
(1 row)

SELECT invalidate_rel_block('tbl_b', 0);
 invalidate_rel_block 
----------------------
 
(1 row)

SELECT invalidate_rel_block('tbl_b', 1);
 invalidate_rel_block 
----------------------
 
(1 row)

SELECT invalidate_rel_block('tbl_b', 2);
 invalidate_rel_block 
----------------------
 
(1 row)

SELECT redact($$
  SELECT count(*) FROM tbl_b;
$$);
NOTICE:  wrapped error: could not read blocks 1..2 in file base/<redacted>: read only 8192 of 16384 bytes
 redact 
--------
 f
(1 row)

SELECT inj_io_short_read_detach();
 inj_io_short_read_detach 
--------------------------
 
(1 row)

-- verify that checksum errors are detected even as part of a shortened
-- multi-block read
-- (tbl_a, block 1 is corrupted)
SELECT redact($$
  SELECT count(*) FROM tbl_a WHERE ctid < '(2, 1)';
$$);
NOTICE:  wrapped error: invalid page in block 2 of relation base/<redacted>
 redact 
--------
 f
(1 row)

SELECT inj_io_short_read_attach(8192);
 inj_io_short_read_attach 
--------------------------
 
(1 row)

SELECT invalidate_rel_block('tbl_a', 0);
 invalidate_rel_block 
----------------------
 
(1 row)

SELECT invalidate_rel_block('tbl_a', 1);
 invalidate_rel_block 
----------------------
 
(1 row)

SELECT invalidate_rel_block('tbl_a', 2);
 invalidate_rel_block 
----------------------
 
(1 row)

SELECT redact($$
  SELECT count(*) FROM tbl_a WHERE ctid < '(2, 1)';
$$);
NOTICE:  wrapped error: invalid page in block 2 of relation base/<redacted>
 redact 
--------
 f
(1 row)

SELECT inj_io_short_read_detach();
 inj_io_short_read_detach 
--------------------------
 
(1 row)

-- FIXME: Should error
-- FIXME: errno encoding?
SELECT inj_io_short_read_attach(-5);
 inj_io_short_read_attach 
--------------------------
 
(1 row)

SELECT invalidate_rel_block('tbl_b', 2);
 invalidate_rel_block 
----------------------
 
(1 row)

SELECT redact($$
  SELECT count(*) FROM tbl_b WHERE ctid = '(2, 1)';
$$);
NOTICE:  wrapped error: could not read blocks 2..3 in file base/<redacted>: Input/output error
 redact 
--------
 f
(1 row)

SELECT inj_io_short_read_detach();
 inj_io_short_read_detach 
--------------------------
 
(1 row)

